runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,261 @@
+package org.corfudb.runtime.collections;
+
+import com.google.protobuf.Message;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.protocols.logprotocol.MultiObjectSMREntry;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.StreamSubscriptionException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+/**
+ * A simple thread based subscription engine where each subscriber or listener gets
+ * a thread which will listen on the tables of interest to it and moves at a rate
+ * which is the same as the consumption.
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private static final int DEFAULT_NUM_SUBSCRIBERS = 6; // Arbitrary sanity check value.
+    private static final int DEFAULT_SLEEP_MILLIS = 50; // Default time to sleep
+    public static final int DEFAULT_LONG_RUNNING_TIME_SECS = 1; // When to log long running onNext()
+    /**
+     * The actual executor thread pool where the subscribers will run.
+     */
+    private ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, StreamSubscriber> subscriptions;
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    @AllArgsConstructor
+    @Getter
+    private class StreamSubscriber {
+        private final SubscriberTask subscriberTask;
+        private final ScheduledFuture<?> scheduledFuture;
+    }
+
+    public StreamManager(@Nonnull CorfuRuntime runtime, int maxSubscribers) {
+        this.runtime = runtime;
+        this.subscriptions = new HashMap<>();
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(maxSubscribers);
```

Comment:
@hisundar I think at least we need to log a warning message if the number of subscribers exceeds limit, otherwise it will be hard to debug when such issues happen. Also is 6 a bit large for default maxSubscribers considering each subscriber will now have a thread?

Reply:
ok, so how about a large limit like 32 at which we will throw an exception. It will only be hit in case there is some bug that is calling subscribe() and leaking?

Reply:
Yea we can do that, but one thing I think that is also important is logging the number of subscribers when there are new subscribes, since the current design is supposed to be used by a small number of subscriptions, so if we hit a performance issue, we can easily tell there are too many subscribers.